File: Editorial/76_Minimum_Window_Substring/C/1_Sliding_Window.c
char* minWindow(char* s, char* t) {
    if (strlen(s) == 0 || strlen(t) == 0) {
        return "";
    }
    // Dictionary which keeps a count of all the unique characters in t.
    int dictT[128] = {0};
    for (int i = 0; i < strlen(t); i++) {
        dictT[t[i]]++;
    }
    // Number of unique characters in t, which need to be present in the desired
    // window.
    int required = 0;
    for (int i = 0; i < 128; i++) {
        if (dictT[i] > 0) required++;
    }
    // Left and Right pointer
    int l = 0, r = 0;
    // formed is used to keep track of how many unique characters in t
    // are present in the current window in its desired frequency.
    // e.g. if t is "AABC" then the window must have two A's, one B and one C.
    // Thus formed would be = 3 when all these conditions are met.
    int formed = 0;
    // Dictionary which keeps a count of all the unique characters in the
    // current window.
    int windowCounts[128] = {0};
    // ans list of the form (window length, left, right)
    int ans[3] = {-1, 0, 0};
    while (r < strlen(s)) {
        // Add one character from the right to the window
        char c = s[r];
        windowCounts[c]++;
        // If the frequency of the current character added equals to the
        // desired count in t then increment the formed count by 1.
        if (dictT[c] != 0 && windowCounts[c] == dictT[c]) {
            formed++;
        }
        // Try and contract the window till the point where it ceases to be
        // 'desirable'.
        while (l <= r && formed == required) {
            c = s[l];
            // Save the smallest window until now.
            if (ans[0] == -1 || r - l + 1 < ans[0]) {
                ans[0] = r - l + 1;
                ans[1] = l;
                ans[2] = r;
            }
            // The character at the position pointed by the
            // `Left` pointer is no longer a part of the window.
            windowCounts[c]--;
            if (dictT[c] != 0 && windowCounts[c] < dictT[c]) {
                formed--;
            }
            // Move the left pointer ahead, this would help to look for a new
            // window.
            l++;
        }
        // Keep expanding the window once we are done contracting.
        r++;
    }
    if (ans[0] == -1) return "";
    char* res = (char*)malloc(sizeof(char) * (ans[0] + 1));
    strncpy(res, s + ans[1], ans[0]);
    res[ans[0]] = '\0';
    return res;
}

File: Editorial/76_Minimum_Window_Substring/TypeScript/1_Sliding_Window.ts
function minWindow(s: string, t: string): string {
    if (s.length === 0 || t.length === 0) return "";
    // Dictionary which keeps a count of all the unique characters in t.
    let dictT: Map<string, number> = new Map();
    for (let i = 0; i < t.length; i++) {
        dictT.set(t.charAt(i), (dictT.get(t.charAt(i)) || 0) + 1);
    }
    // Number of unique characters in t, which need to be present in the desired window.
    let required: number = dictT.size;
    // Left and Right pointer
    let l = 0,
        r = 0;
    // formed is used to track how many unique characters in the desired window.
    let formed: number = 0;
    // Dictionary which keeps a count of all the unique characters in the window.
    let windowCounts: Map<string, number> = new Map();
    // ans list of the form (window length, left, right)
    let ans: number[] = [-1, 0, 0];
    while (r < s.length) {
        // Add one character from the right into window.
        let c: string = s.charAt(r);
        windowCounts.set(c, (windowCounts.get(c) || 0) + 1);
        // If the frequency of the current character added equals the count, increase `forms` counter.
        if (dictT.has(c) && windowCounts.get(c) === dictT.get(c)) {
            formed++;
        }
        while (l <= r && formed === required) {
            c = s.charAt(l);
            // Save the smallest window.
            if (ans[0] === -1 || r - l + 1 < ans[0]) {
                ans[0] = r - l + 1;
                ans[1] = l;
                ans[2] = r;
            }
            // Move the left pointer forward to explore a new window.
            windowCounts.set(c, windowCounts.get(c)! - 1);
            if (dictT.has(c) && windowCounts.get(c)! < dictT.get(c)!) {
                formed--;
            }
            l++;
        }
        // Keep expanding the window once we are done contracting.
        r++;
    }
    return ans[0] === -1 ? "" : s.substring(ans[1], ans[2] + 1);
}
